### PR TITLE
perf improvements in tslibs.period

### PR DESCRIPTION
**Pull request by jbrockmendel**
Removes an unnecessary call to `frequencies.to_offset`, one of the last few non-cython dependencies in the file.

will post asv results when available.

### CODE REVIEW

File: pandas/_libs/tslibs/period.pyx [hunk 1]
```diff
@@ -1058,18 +1058,20 @@ cdef class _Period(object):
         return hash((self.ordinal, self.freqstr))
 
     def _add_delta(self, other):
-        if isinstance(other, (timedelta, np.timedelta64, offsets.Tick)):
-            offset = frequencies.to_offset(self.freq.rule_code)
-            if isinstance(offset, offsets.Tick):
```
Comment (jbrockmendel) [inline pandas/_libs/tslibs/period.pyx]: This along with the line above is redundant; `isinstance(offset, offsets.Tick)` if and only if `isinstance(self.freq, offsets.Tick)`.

**Comment by gfyoung** [comment]
> will post asv results when available.

@jbrockmendel : `whatsnew` entry in `0.24.0` + perf test would also be good here.

File: pandas/_libs/tslibs/period.pyx [hunk 2]
```diff
@@ -1058,18 +1059,21 @@ cdef class _Period(object):
...
                 nanos = delta_to_nanoseconds(other)
                 offset_nanos = delta_to_nanoseconds(offset)
-
                 if nanos % offset_nanos == 0:
                     ordinal = self.ordinal + (nanos // offset_nanos)
                     return Period(ordinal=ordinal, freq=self.freq)
             msg = 'Input cannot be converted to Period(freq={0})'
             raise IncompatibleFrequency(msg.format(self.freqstr))
-        elif isinstance(other, offsets.DateOffset):
+        elif getattr(other, "_typ", None) == "dateoffset":
```
Comment (jreback) [inline pandas/_libs/tslibs/period.pyx]: maybe should add an is_offset_object in util.pxd

**Review by jreback — CHANGES_REQUESTED** [review]
+ @gfyoung comments. (just see if we have an asv, you can run them and run report and significant diffs)

**Comment by jbrockmendel** [comment]
as isn't showing much (recall it is wonky on my machine for reasons unknown)

```
taskset 6 asv continuous -f 1.1 -E virtualenv master HEAD -b period
[...]
    before     after       ratio
  [defdb34b] [daf9aaf7]
-    3.26ms     2.82ms      0.87  period.PeriodIndexConstructor.time_from_pydatetime('D')

    before     after       ratio
  [defdb34b] [daf9aaf7]
+   87.34ms   147.38ms      1.69  gil.ParallelDatetimeFields.time_datetime_to_period
+  118.47ms   133.50ms      1.13  gil.ParallelDatetimeFields.time_period_to_datetime
+   35.94ms    39.66ms      1.10  period.DataFramePeriodColumn.time_setitem_period_column
-   66.75μs    60.54μs      0.91  period.PeriodProperties.time_property('min', 'start_time')
-   75.69μs    68.21μs      0.90  period.PeriodProperties.time_property('M', 'end_time')

    before     after       ratio
  [defdb34b] [daf9aaf7]
+  884.44ns     1.00μs      1.13  period.PeriodProperties.time_property('min', 'week')
-  143.83ms   116.09ms      0.81  gil.ParallelDatetimeFields.time_period_to_datetime
-  143.24ms    93.98ms      0.66  gil.ParallelDatetimeFields.time_datetime_to_period

    before     after       ratio
  [defdb34b] [daf9aaf7]
+  123.27ms   164.93ms      1.34  gil.ParallelDatetimeFields.time_period_to_datetime
-   19.73μs    17.71μs      0.90  period.PeriodUnaryMethods.time_asfreq('M')
-    1.03μs   879.73ns      0.85  period.PeriodProperties.time_property('M', 'is_leap_year')

    before     after       ratio
  [defdb34b] [daf9aaf7]
+  112.56ms   125.62ms      1.12  gil.ParallelDatetimeFields.time_datetime_to_period
-   40.44ms    36.20ms      0.89  period.DataFramePeriodColumn.time_setitem_period_column
```

File: pandas/_libs/tslibs/period.pyx (hunk 2 above)
Comment (jreback) [reply pandas/_libs/tslibs/period.pyx]: can you do this

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks!

**Pull request merged**